### PR TITLE
Update shinytest2 for widgets

### LINKED ISSUE #296: Implement regression test
### Feature description

Related with https://github.com/insightsengineering/coredev-tasks/issues/589

### Scope of work

teal.slice (100 hours)
- data type (7 data types)
- data (4 data objects, MAE/SE and DF/Matrix)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct.

### Contribution Guidelines

- [x] I agree to follow this project's Contribution Guidelines.

### Security Policy

- [x] I agree to follow this project's Security Policy.

### PR DESCRIPTION
**Pull request by llrs-roche**
# Pull Request

Fixes #296

While working on the tests I changed `rvest::html_nodes` to `rvest::html_element`. 
The former was deprecated in rvest 1.0.0 some time ago. 

The issue list 8 widgets but I only found 3 that have ui/server. 

- [ ] `plot_with_settings()`: Missing testing scrollbar functionality
- [x] `table_with_settings()`
- [x] `verbatim_popup()`
- [ ] `get_dt_rows()`: No previous tests.
- [ ] `optionalSelectInput()`/`updateOptionalSelectInput()`: No previous tests
- [ ] `draggable_buckets()`: no scrollbar, no previous tests.
- [ ] `nested_closeable_modal()`: No previous tests
- [ ]  `optionalSliderInputValMinMax()`: one of the examples isn't displayed: 

<details><summary>Details</summary>
<p>

The app is open but no slider is shown
```r
ui <- bslib::page_fluid(
    shinyjs::useShinyjs(),
    optionalSliderInputValMinMax("a1", "b1", 1)
)
if (interactive()) {
    shiny::shinyApp(ui, function(input, output) {})
}
```

</p>
</details>

### CODE REVIEW

File: tests/testthat/test-draggable_buckets_ui.R [hunk 1]
```diff
@@ -0,0 +1,24 @@
+app_driver_db <- function() {
+  ui <- bslib::page_fluid(
```
[the remark this answers is not in the document]
Comment (averissimo) [reply tests/testthat/test-draggable_buckets_ui.R:2]: alternatively, use this at the top of the files so the whole file can be skipped

File: tests/testthat/test-get_dt_rows_ui.R [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    server = function(input, output, session) server("my_table_module")
+  )
+}
+
+
+testthat::test_that(
+  "e2e: teal.widgets::get_dt_rows: table is visible",
+  {
+    skip_if_too_deep(5)
+    testthat::skip_if_not_installed("DT")
```
Comment (averissimo) [inline tests/testthat/test-get_dt_rows_ui.R:40]: No need for this here

```suggestion
```

File: tests/testthat/test-get_dt_rows_ui.R [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    app_driver <- shinytest2::AppDriver$new(
+      app_driver_gdr(),
+      name = "gdr",
+      variant = "app_driver_gdr_ui"
+    )
```
Comment (averissimo) [inline tests/testthat/test-get_dt_rows_ui.R:45]: This app fails with:

```
app_driver$get_logs()
#> {shiny}      R  stderr ----------- Loading required package: shiny
#> {shiny}      R  stderr ----------- Running application in test mode.
#> {shiny}      R  stderr ----------- 
#> {shiny}      R  stderr ----------- Listening on http://127.0.0.1:5354
#> {shiny}      R  stderr ----------- `shiny::renderDataTable()` is deprecated as of shiny 1.8.1.
#> {shiny}      R  stderr ----------- Please use `DT::renderDT()` instead.
#> {shiny}      R  stderr ----------- Since you have a suitable version of DT (>= v0.32.1), shiny::renderDataTable() will automatically use DT::renderDT() under-the-hood.
#> {shiny}      R  stderr ----------- If this happens to break your app, set `options(shiny.legacy.datatable = TRUE)` to get the legacy datatable implementation (or `FALSE` to squelch this message).
#> {shiny}      R  stderr ----------- See <https://rstudio.github.io/DT/shiny.html> for more information.
#> {shiny}      R  stderr ----------- Warning: Error in $: Can't access reactive value 'dt_rows' outside of reactive consumer.
#> {shiny}      R  stderr ----------- ℹ Do you need to wrap inside reactive() or observe()?
#> {shiny}      R  stderr -----------   67: <Anonymous>
#> {shiny}      R  stderr -----------   66: signalCondition
#> {shiny}      R  stderr -----------   65: signal_abort [/tmp/RtmpFSijkY/R.INSTALL1d831510d7c59/rlang/R/cnd-abort.R#884]
#> {shiny}      R  stderr -----------   64: rlang::abort [/tmp/RtmpFSijkY/R.INSTALL1d831510d7c59/rlang/R/cnd-abort.R#400]
#> {shiny}      R  stderr -----------   63: $.reactivevalues [/tmp/RtmphsUIme/R.INSTALL2a77d6efab65e/shiny/R/reactives.R#635]
#> {shiny}      R  stderr -----------   59: module [#14]
#> {shiny}      R  stderr -----------   54: callModule [/tmp/RtmphsUIme/R.INSTALL2a77d6efab65e/shiny/R/modules.R#192]
#> {shiny}      R  stderr -----------   53: moduleServer [/tmp/RtmphsUIme/R.INSTALL2a77d6efab65e/shiny/R/modules.R#165]
#> {shiny}      R  stderr -----------   52: server [#13]
#> {shiny}      R  stderr -----------   51: server [#23]
#> {shiny}      R  stderr -----------   14: <Anonymous> [/tmp/RtmphsUIme/R.INSTALL2a77d6efab65e/shiny/R/runapp.R#388]
#> {shiny}      R  stderr -----------   12: <Anonymous>
#> {shiny}      R  stderr -----------   11: <Anonymous>
#> {shiny}      R  stderr -----------   10: base::do.call
#> {shiny}      R  stderr -----------    9: base::saveRDS
#> {shiny}      R  stderr -----------    8: base::withCallingHandlers
#> {shiny}      R  stderr -----------    7: doTryCatch
#> {shiny}      R  stderr -----------    6: tryCatchOne
#> {shiny}      R  stderr -----------    5: tryCatchList
#> {shiny}      R  stderr -----------    2: tryCatchList
#> {shiny}      R  stderr -----------    1: base::tryCatch
#> {shiny}      R  stderr ----------- Error in input$dt_rows : 
#> {shiny}      R  stderr -----------   Can't access reactive value 'dt_rows' outside of reactive consumer.
#> {shiny}      R  stderr ----------- ℹ Do you need to wrap inside reactive() or observe()?
```

Comment (averissimo) [reply tests/testthat/test-get_dt_rows_ui.R:45]: You probably want to use the following in the shinyApp (and probably update the example)

```r
output$data_table <- DT::renderDataTable(
  DT::datatable(iris, options = list(pageLength = input$dt_rows))
)
```

File: tests/testthat/test-get_dt_rows_ui.R [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    app_driver$wait_for_idle(timeout = default_idle_timeout)
+    testthat::expect_true(is_visible("#my_table_module-data_table", app_driver))
```
Comment (averissimo) [inline tests/testthat/test-get_dt_rows_ui.R:47]: This test doesn't do anything other than test `DT` and even so it fails to detect if it was rendered.

Do you want to test the change of value via `Shiny.onInputChange`?

File: tests/testthat/test-nested_closeable_modal_ui.R [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      tags$div("one modal")
+    )
+  )
+  shiny::shinyApp(ui, function(input, output) {
+    observeEvent(input$show_1, {
+      shinyjs::runjs("$(\"#modal_1\").modal(\"show\")")
+    })
+  })
+}
+click_modal <- "document.querySelector('#show_1').click();"
```
Comment (averissimo) [inline tests/testthat/test-nested_closeable_modal_ui.R:22]: Why do you need this?

File: tests/testthat/test-nested_closeable_modal_ui.R [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      app_driver_ncm(),
+      name = "ncm",
+      variant = "app_driver_ncm_ui"
+    )
+    app_driver$wait_for_idle(timeout = default_idle_timeout)
+    testthat::expect_true(is_visible("#show_1", app_driver))
+    testthat::expect_false(is_visible(".modal-content", app_driver))
+    app_driver$stop()
+  }
+)
```
Comment (averissimo) [inline tests/testthat/test-nested_closeable_modal_ui.R:38]: Add a test that popups the modal? otherwise we are not testing interactivity here

File: tests/testthat/test-optionalSelectInput_ui.R [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+      variant = "app_driver_osi_ui"
+    )
+    app_driver$wait_for_idle(timeout = default_idle_timeout)
+    selector <- sprintf(
+      "body > div.container-fluid > bslib-layout-columns > div:nth-child(%s)",
+      1:6
+    )
+    for (sel in selector) {
+      testthat::expect_true(is_visible(sel, app_driver))
+    }
```
Comment (averissimo) [inline tests/testthat/test-optionalSelectInput_ui.R:108]: Maybe check if outputs are the expected?

Or.. this might be a good candidate to screenshot, given that there are a bunch of elements

image

**Review by averissimo — CHANGES_REQUESTED** [review]
One of the tests has a failure on my machine. 

The test passes as the element is visible, but completely empty.

In some cases where the visibility of the element is the feature I think you can use `is_visible`, but for others consider checking for element content, its output or other clear indication of what is expected.

File: tests/testthat/test-get_dt_rows_ui.R (hunk 4 above)
Comment (llrs-roche) [reply tests/testthat/test-get_dt_rows_ui.R:47]: Indeed, I updated to check that DT is displayed.

Regarding testing, I indeed didn't test what this widget does. But I have some problems doing it, see the [documented limitation](https://rstudio.github.io/shinytest2/articles/in-depth.html#limitations) of shinytest2:

> These include some [htmlwidgets](http://www.htmlwidgets.org/), such as [DT](https://rstudio.github.io/DT/) and [plotly](https://github.com/plotly/plotly.R), as well as Shiny’s built-in plot interactions with [renderPlot()](https://rdrr.io/pkg/shiny/man/renderPlot.html).
> However, the browser will not be able to tell the input object to do the exact same behaviors. 
> For components that have internal state that is updated in response to user interaction, that internal state will not be updated when the test script is played. In some cases, this may mean that when the script is played, the behavior when the script is played will not be the same as when a user actually interacts with the application.

I haven't been able to activate the rows change or even extract the current number of rows.  With:

```r
app_driver$set_inputs(`my_table_module-data_table_state-length` = 25, allow_no_input_binding_ = TRUE)
```
Or using `app_driver$run_js`.
I've set an expectation to check the default number of rows, but without changing the number of rows the UI isn't updated so I cannot test that it actually reflects the change on shinytest2. 
Let me know if you have some suggestions to test this.

**Review by llrs-roche — COMMENTED** [review]
Thanks for this second review. It has uncovered quite a few issues.
There is one I couldn't completely address about `get_dt_rows`. Let me know if there is something I missed.

File: tests/testthat/test-get_dt_rows_ui.R (hunk 4 above)
Comment (averissimo) [reply tests/testthat/test-get_dt_rows_ui.R:39]: I have a suggestion that takes advantage of DT itself by adding a button that the change and re-render.

```diff
diff --git a/tests/testthat/test-get_dt_rows_ui.R b/tests/testthat/test-get_dt_rows_ui.R
index 092af8aa..a730fc6f 100644
--- a/tests/testthat/test-get_dt_rows_ui.R
+++ b/tests/testthat/test-get_dt_rows_ui.R
@@ -12,11 +12,28 @@ app_driver_gdr <- function() {
   # use the input$dt_rows in the Shiny Server function
   server <- function(id) {
     moduleServer(id, function(input, output, session) {
-      output$data_table <- DT::renderDataTable(iris)
+      output$data_table <- DT::renderDataTable(
+        iris,
+        extensions = 'Buttons',
+        options = list(
+          dom = 'Bfrtip',
+          buttons = list(
+            list(
+              extend = "collection",
+              text = 'Show 25',
+              attr = list(id = 'show_25'),
+              action = DT::JS(
+              "function ( e, dt, node, config ) { dt.page.len(25); dt.ajax.reload(); }")
+            )
+          )
+        )
+      )
       rows <- reactive({
         paste0("Selected Rows ", input$dt_rows)
       })
       output$rows <- renderText(rows())
+
+      shiny::exportTestValues(table_proxy = DT::dataTableProxy('data_table'))
     })
   }
 
@@ -43,11 +60,12 @@ testthat::test_that(
     testthat::expect_equal(dt_state$input$`my_table_module-data_table_state`$length, 10)
 
     # Change rows value
-    # FIXME: Couldn't find a way to programatically change that
+    app_driver$click(selector = "#show_25")
 
     # Check value again
     dt_state <- app_driver$get_values(input = "my_table_module-data_table_state")
-    testthat::expect_equal(dt_state$input$`my_table_module-data_table_state`$length, 10)
+    testthat::expect_equal(dt_state$input$`my_table_module-data_table_state`$length, 25)
+    testthat::expect_equal(app_driver$get_value(input = "my_table_module-dt_rows"), 25)
 
     app_driver$stop()
   }
```

**Review by averissimo — APPROVED** [review]
Great job addresing all of the comments.

I left a suggestion on the limitation with DT. I believe it works, but it wasn't easy to find :see_no_evil: 

Once that's addressed feel free to merge the PR

:100:
